### PR TITLE
Time complexity comments

### PR DESCRIPTION
**Pull request by canadianschism**
So I think I've got the Tc stuff down. It took longer than I expected because I'm still not 100% sure on how the whole thing works. I'm really not sure about the merge sort algo. From what I've been reading online:

- Splitting is O(log n) <- this would be the getMiddleNode part, as it's the splitter using the fast/slow search
- The mergeSort is O(n)
- The sortedMerge is O(n log n)

So, by that logic it would be O(log n), however what I'm seeing is that mergeSort on linked lists are O(n log n) ... So, yeah. If we can get together to discuss this I'd feel better about things.

### CODE REVIEW

**Comment by richieyoum** [comment]
O(log n) < O(n) < O(n log n). So the time complexity of the algorithm should be O(n log n).

**Comment by canadianschism** [comment]
Alright, I've changed things around, along with the answer. As he asked for the answer to be in the main method I've moved it down there, to be O(n log n). I'm not sure if I have to put in another pull request, or if pushes that I'm doing now will automatically be added?

**Comment by richieyoum** [comment]
You can simply push it on your branch and it'll reflect in the current PR. Thanks!

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 1]
```diff
@@ -16,34 +18,34 @@ public SinglyLinkedList(){
...
+    	// Tc = 1c
         return size;
     }
 
     /**
-     * Checks whether linked list is empty
      * @return boolean indicating whether list is empty
      */
     public boolean isEmpty(){
+    	// Tc = 1ct
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:32]:
```suggestion
    	// Tc = 1c
```
I think you made a typo?

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 2]
```diff
@@ -16,34 +18,34 @@ public SinglyLinkedList(){
 
     // Access methods
     /**
-     * Returns number of elements in linked list
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:19]: I know it's kind of duplicate info, but javadoc usually includes the descriptions on top of any parameters + return variables. Could you add them back in please? :)

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 3]
```diff
@@ -53,17 +55,22 @@ public Node getTailNode(){
      * @param name new bird name
      */
     public void addHeadNode(String name, int count){
+    	// addHeadNode: Tc = 11c
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:58]: shouldn't it be 7c?

line 61 - 2c
line 64 - 1c
line 67 - 1c
line 70 - 1c
line 71 - 0 to 1c
line 74 - 1c

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 4]
```diff
@@ -72,29 +79,32 @@ public void addHeadNode(String name, int count){
      * @param name new bird name
      */
     public void addTailNode(String name, int count){
+    	// addTailNode: Tc = 12c
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:82]: same comment as the one in addHeadNode

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 5]
```diff
@@ -72,29 +79,32 @@ public void addHeadNode(String name, int count){
...
-        // increment the size
+        // increment the size of the list
         size++;
     }
 
     /**
      * Remove the head node
      */
     public void removeHeadNode(){
+    	// removeHeadNode: Tc = 7c
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:104]: I think it's 4c

106 - 1c
108 - 1c
110 - 0 ~ 1c
111 - 0 ~ 1c

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 6]
```diff
@@ -1,11 +1,13 @@
 package assignment1;
 
+
 import java.io.BufferedReader;
 import java.io.File;
 import java.io.FileReader;
 import java.io.IOException;
 
 public class SinglyLinkedList {
+	// 3 operations, constant
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:10]: do we need to count each constant for the assignment? or can we just leave that it's constant, linear ie- O(n), etc...?

Comment (canadianschism) [reply assignment1/src/assignment1/SinglyLinkedList.java]: I'm doing counts for each method, and then resolving in the main method. I've changed that line to Tc = 3c 

But to be honest, I have zero idea.

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 7]
```diff
@@ -207,11 +225,12 @@ private Node getMiddleNode(Node tempHead){
 
     /**
      * Recursive calls to sort all items between the two linked list, and then merge them together
-     * @param left a node to be compared with right node
-     * @param right a node to be compared with left node
-     * @return merged node from left and right nodes in ascending order
+     * @param left		: a node to be compared with right node
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:228]: what's up with the indent? :thinking:

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 8]
```diff
@@ -263,6 +283,7 @@ private Node mergeSort(Node tempHead){
      * Merge sort the current linked list
      */
     public void sortList(){
+    	// sortList: 3c
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:286]: 1c

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 9]
```diff
@@ -271,6 +292,8 @@ public void sortList(){
      * Print the linked list items
      */
     public void printNodeItems(){
+    	// printNodeItems: 3c + 0(n)
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:295]: let's jsut abstract it as constant, so just keep O(n)

File: assignment1/src/assignment1/SinglyLinkedList.java (hunk 7 above)
Comment (canadianschism) [reply assignment1/src/assignment1/SinglyLinkedList.java:228]: I have this thing with stuff lining up on the page. I'll put that back. It was bothering the bejeesus out of me.

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 10]
```diff
@@ -341,17 +371,31 @@ public int getCount(){
...
         } catch (IOException e){
+        	// if there is a problem reading the file, print out the line number/method
+        	// where the error occured for debugging
             e.printStackTrace();
         }
         return null;
     }
 
     public static void main(String[] args) throws IOException{
+    	/*	Count of method operations in main method
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:388]: I can't confirm the individual constants, but it is O(n log n). Please update after reviewing my comments above.

File: assignment1/src/assignment1/SinglyLinkedList.java (hunk 3 above)
Comment (canadianschism) [reply assignment1/src/assignment1/SinglyLinkedList.java:58]: I was thinking that sending something in a method call was 1c ... so it should actually be 12.

Still can't find anything related to method calls and the like, but I remember him saying something like "get the value of the variable" being 1 operation. In my head, calling a new Node(name, count) means that, at the very least, you're getting the value of name and count (2c), and setting it to newNode (1c). I think I'm going to email him about that.

However, at this point, I'm wondering if we shouldn't just be using O(1) for methods whose Tc are constant.

"Comments: in the comments for every method, you have to **specify the time complexity of that method**, and finally for the main method, specify the time complexity of the whole program."

Counting vs. O(1) comes to the same thing, wouldn't it? Your thoughts?

File: assignment1/src/assignment1/SinglyLinkedList.java (hunk 9 above)
Comment (canadianschism) [reply assignment1/src/assignment1/SinglyLinkedList.java]: Done!

File: assignment1/src/assignment1/SinglyLinkedList.java (hunk 10 above)
Comment (canadianschism) [reply assignment1/src/assignment1/SinglyLinkedList.java:392]: Done, I've reviewed, though I thought I'd had O(n log n) in there originally. In any case, reviewed and corrected!

File: assignment1/src/assignment1/SinglyLinkedList.java (hunk 8 above)
Comment (canadianschism) [reply assignment1/src/assignment1/SinglyLinkedList.java]: Resolved to O(1) as per Amin

File: assignment1/src/assignment1/SinglyLinkedList.java (hunk 6 above)
Comment (canadianschism) [reply assignment1/src/assignment1/SinglyLinkedList.java]: This has been taken care of, with each method that is constant being O(1)

File: assignment1/src/assignment1/SinglyLinkedList.java (hunk 5 above)
Comment (canadianschism) [reply assignment1/src/assignment1/SinglyLinkedList.java]: Brought down to O(1)

File: assignment1/src/assignment1/SinglyLinkedList.java (hunk 2 above)
Comment (canadianschism) [reply assignment1/src/assignment1/SinglyLinkedList.java]: I apologize for this... again, not a programmer, so I didn't know :) I think it should all be back to the way it was.

File: assignment1/src/assignment1/SinglyLinkedList.java [hunk 11]
```diff
@@ -36,14 +40,16 @@ public boolean isEmpty(){
      * @return list head node
      */
     public Node getHeadNode(){
+    	// Tc = O(1)
         return head;
     }
 
     /**
-     * Returns list tail node
+     * Returns list tal node
```
Comment (richieyoum) [inline assignment1/src/assignment1/SinglyLinkedList.java:48]: tail*

**Comment by richieyoum** [comment]
LGTM! Merging now.

**Pull request merged**